liquid-compiler/src/parser.rs

```diff
@@ -175,31 +176,60 @@ fn parse_value(value: Pair) -> Expression {
 
 /// Parses a `FilterCall` from a `Pair` with a filter.
 /// This `Pair` must be `Rule::Filter`.
-fn parse_filter(filter: Pair, options: &Language) -> Result<FilterCall> {
+fn parse_filter(filter: Pair, options: &Language) -> Result<Box<Filter>> {
     if filter.as_rule() != Rule::Filter {
         panic!("Expected a filter.");
     }
 
     let mut filter = filter.into_inner();
     let name = filter.next().expect("A filter always has a name.").as_str();
-    let args = filter.map(parse_value).collect();
-
-    let f = options
-        .filters
-        .get(name)
-        .ok_or_else(|| {
-            let mut available: Vec<_> = options.filters.plugin_names().collect();
-            available.sort_unstable();
-            let available = itertools::join(available, ", ");
-            Error::with_msg("Unknown filter")
-                .context("requested filter", name.to_owned())
-                .context("available filters", available)
-        })?
-        .clone();
-    let f = FilterCall::new(name, f, args);
+
+    let mut keyword_args = Vec::new();
+    let mut positional_args = Vec::new();
+
+    for arg in filter {
+        match arg.as_rule() {
+            Rule::PositionalFilterArgument => {
+                let value = arg.into_inner().next().expect("Rule ensures value.");
+                let value = parse_value(value);
+                positional_args.push(value);
+            }
+            Rule::KeywordFilterArgument => {
+                let mut arg = arg.into_inner();
+                let key = arg.next().expect("Rule ensures identifier.").as_str();
+                let value = arg.next().expect("Rule ensures value.");
+                let value = parse_value(value);
+                keyword_args.push((key, value));
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    let args = FilterArguments {
+        positional: Box::new(positional_args.into_iter()),
+        keyword: Box::new(keyword_args.into_iter()),
+    };
+
+    let f = options.filters.get(name).ok_or_else(|| {
+        let mut available: Vec<_> = options.filters.plugin_names().collect();
+        available.sort_unstable();
+        let available = itertools::join(available, ", ");
+        Error::with_msg("Unknown filter")
+            .context("requested filter", name.to_owned())
+            .context("available filters", available)
+    })?;
+    let f = f.parse(args)?;
+
     Ok(f)
 }
 
+/// Structure that holds both the positional and the keyword arguments of
+/// a filter.
+pub struct FilterArguments<'a> {
```

Comment:
Should this be in with the filter traits?

Reply:
The thing is, right now I'm working on the liquid::error messages (i think it is the only actual feature i didn't completely finish, the rest will be tests and code cleanup)

In order to improve the liquid error messages, I will be sending the span of the filter as a private member of this struct

```
pub struct FilterArguments<'a> {
    pub positional: Box<Iterator<Item = Expression>>,
    pub keyword: Box<Iterator<Item = (&'a str, Expression)> + 'a>,
    span: Span<'a>
}
```

With methods that will allow to raise errors with the line/collumn information. However, putting this with the filter traits would mean using pest types outside `parser.rs`. I personally thing it would be best that everything pest-related should stay in that file (and if it gets too big, we could eventually split it into submodules).

Reply:
A couple of questions
- Should we be reworking the errors in this PR?
  - My suspicion is to hold off on doing it this PR (1) to keep this PR smaller (its already huge) and (2) so we can start prototyping and iterating on our error solution in less impactful ways, API wise.
- Should we be using `Span` for this?
  - Long term, we want at least a subset of this information available during `render` and not just `parse` which means we'll be needing to move away from `Span` to another way of tracking this state.

Reply:
- Ok, with that in mind I will remove this part from the PR
  - My initial intention was to compensate the regression on the error messages that this PR eventually carries (however the current error messages aren't really that good anyway)
- Well, I think we will end up needing some span-like structure in the end regardless of how we implement it.
  - We could easily switch from `Span` in the future, because it was a private member.